codebase2/codebase-sqlite/U/Codebase/Sqlite/Queries.hs

```diff
@@ -1408,6 +1409,108 @@ getDependencyIdsForDependent dependent@(C.Reference.Id oid0 _) =
     isNotSelfReference (C.Reference.Id oid1 _) =
       oid0 /= oid1
 
+-- | Given two term (components) A and B, return the set of all terms that are along any "dependency path" from A to B,
+-- not including A nor B; i.e., the transitive dependencies of A that are transitive dependents of B.
+--
+-- For example, if A depends on X and Y, and Y depends on Z, and X and Z depend on B...
+--
+--     --X--
+--    /     \
+--   A       B
+--    \     /
+--     Y---Z
+--
+-- ...then `getDependenciesBetweenTerms A B` would return the set {X Y Z}
+getDependenciesBetweenTerms :: ObjectId -> ObjectId -> Transaction (Set ObjectId)
+getDependenciesBetweenTerms oid1 oid2 =
+  queryListCol sql (oid1, oid2, oid2) <&> Set.fromList
+  where
+    -- Given the example above, we'd have tables that look like this.
+    --
+    -- First, the `paths` table finds all paths from source `A`, exploring depth-first. As a minor optimization, we seed
+    -- the search not with `A`, but rather the direct dependencies of `A` (namely `X` and `Y`).
+    --
+    -- +-paths-------------------------------+
+    -- +-level-+-object_id-+-object_id_array-+
+    -- |     0 |         X |              '' |
+    -- |     0 |         Y |              '' |
+    -- |     1 |         B |            'X,' |
+    -- |     1 |         Z |            'Y,' |
+    -- |     2 |         B |          'Z,Y,' |
+    -- +-------+-----------+-----------------+
+    --
+    -- Next, we seed another recursive CTE with those paths that end in the sink `B`. This is just the (very verbose)
+    -- way to unnest an array in SQLite. All we're doing is turning the set of strings {'X,' 'Z,Y,'}, each of which
+    -- represents the inner nodes of a full path between `A` and `B`, into the set {X Z Y}, which is just the full set
+    -- of such inner nodes, along any path.
+    --
+    -- +-elems-----------------------+
+    -- +-object_id-+-object_id_array-+
+    -- |           |            'X,' |
+    -- |           |          'Z,Y,' |
+    -- |       'X' |              '' |
+    -- |       'Z' |            'Y,' |
+    -- |       'Y' |              '' |
+    -- +-----------+-----------------+
+    --
+    -- And finally, we just select out the non-null `object_id` rows from here, casting the strings back to integers for
+    -- clarity (this isn't very matter - SQLite would cast on-the-fly).
+    --
+    -- +-object_id-+
+    -- |         X |
+    -- |         Z |
+    -- |         Y |
+    -- +-----------+
+    --
+    -- Notes
+    --
+    -- (1) We only care about term dependencies, not type dependencies
```

Comment:
Why is that? Is it not possible to end up creating a new cycle with types for some reason?

Reply:
Yeah, in a later PR when we remember what propagating changes through types should mean — maybe requiring kind checking, too. In the end, I imagine it will look similar to this PR though.